File: AlgorithmDataStructure/LeetCode/1248-NumOfSubarrays/subarray.c


#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 nums = [2,2,2,1,2,2,1,2,2,2], k = 2
 ans = 16 , [1,2,2,1] 前面四个2 后面四个2，4*4=16
 
 [-1,3,6,10,0,0,0,0,0,0,0,0]
 (3-(-1))*(10-6) = 16
 */
int numberOfSubarrays(int* nums, int numsSize, int k){
    if (nums == NULL || k <= 0 || numsSize < k) { return 0; }
    
    int *odd = (int *)malloc(sizeof(int) * (numsSize + 2));
    memset(odd, 0, sizeof(int) * (numsSize + 2));
    
    int ans = 0, count = 0;
    
    //记录奇数的下标
    for (int i = 0; i < numsSize; i++) {
        if (nums[i] & 1) { odd[++count] = i; }
    }
    //注意边界的处理
    odd[0] = -1; odd[++count] = numsSize;
    
    for (int i = 1; i + k <= count; i++) {
        ans += (odd[i] - odd[i - 1]) * (odd[i + k] - odd[i + k - 1]);
    }
    return ans;
}

int main(int argc, const char * argv[]) {
    
    int a[] = {2,2,2,1,2,2,1,2,2,2};
    printf("%d \n", numberOfSubarrays(a, 10, 2));
    
    return 0;
}

File: AlgorithmDataStructure/LeetCode/1248-NumOfSubarrays/subarray.swift


import Foundation

class Solution {
    
    func numberOfSubarrays(_ nums: [Int], _ k: Int) -> Int {
        let n = nums.count
        if n <= 0 || n < k { return 0}
        
        var count = Array(repeating: 0, count: n + 1)
        var odd = 0, ans = 0
        
        count[0] = 1
        for i in 0 ..< n {
            odd += nums[i] & 1
            print(odd)
            ans += odd >= k ? count[odd - k] : 0
            print(ans, count)
            count[odd] += 1
            print(count)
            print("----------------------")
        }
        
        return ans
    }
}

let s = Solution()
print(s.numberOfSubarrays([2,2,2,1,2,2,1,2,2,2], 2))
